### PR TITLE
Add `HTTPXMock` testing and drop Python 3.8 support

### PR DESCRIPTION
**Pull request by GitOldGrumpy**
This adds a series of tests using httpx_mock and static data files. The work to add the tests has identified two issues:
- During authentication the response from sending the username was not checked correctly. It will actually tell you if the username is invalid.
- For some reason we were always requested the trips for the current day when requesting the static data. I'm not sure why.

I could break these two issues out into multiple pull requests but for the minute they are contained in this pull request.

The merge looks large but this is to do with all the new static JSON files.

### CODE REVIEW

**Comment by GitOldGrumpy** [comment]
@CM000n I cant add the required package(pytest-httpx) to poetry as it requires python >=3.9 according to poetry. I know it works fine on `Python 3.8.18 (default, Aug 25 2023, 13:20:30)` as that is my development environment. Is there a way around this?

**Comment by CM000n** [comment]
> @CM000n I cant add the required package(pytest-httpx) to poetry as it requires python >=3.9 according to poetry. I know it works fine on `Python 3.8.18 (default, Aug 25 2023, 13:20:30)` as that is my development environment. Is there a way around this?

Unfortunately not. If these are the minimum requirements specified for `pytest-httpx`, there is no sensible way around this.
Of course, we could also drop support for Python 3.8. But this is a fundamental decision that should also be approved by @DurgNomis-drol.

**Comment by DurgNomis-drol** [comment]
Please divide you PR into smaller ones, it makes it a lot easier, and faster to approve them ❤️

**Comment by DurgNomis-drol** [comment]
@CM000n I believe @joro75 is depended on 3.8

**Comment by GitOldGrumpy** [comment]
If you look at the code changes they are fairly small, the reason that the pull request is so large is because of the JSON files that provide the data for the tests.

**Comment by DurgNomis-drol** [comment]
@GitOldGrumpy  Personally I would prefer if we could remove support for .8 and .9, but @joro75 is supporting it. So I think to be fair, he should be involved in the decision

**Comment by CM000n** [comment]
So let's hope for a prompt response from @joro75. But I would also prefer to completely discontinue support for 3.8 rather than simply disconnect the tests for it.

**Comment by CM000n** [comment]
Oh dear, it looks like Domoticz, for which @joro75 maintains the Toyota plugin, is not exactly up to date as far as the python versions are concerned: https://www.domoticz.com/wiki/Using_Python_plugins#Required:_install_Python 🙄 
So we will probably have to maintain support for python 3.9 for the time being @DurgNomis-drol.

File: pyproject.toml [hunk 1]
```diff
@@ -27,14 +27,15 @@ repository = "https://github.com/DurgNomis-drol/mytoyota"
...
-python = "^3.8.1"
+python = "^3.9"
 langcodes = "^3.1"
 httpx = ">=0.18.1"
 arrow = "^1.1"
 importlib-metadata = "^7.0.1"
 pyjwt = "^2.8.0"
 pydantic = "^1.10.0"
 hishel = "^0.0.19"
+pytest-httpx = "^0.28.0"
```
Comment (CM000n) [inline pyproject.toml:38]: This should go under dev-dependencies 😉

**Review by CM000n — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (GitOldGrumpy) [reply pyproject.toml:38]: poetry & toml is all new to me!

Comment (CM000n) [reply pyproject.toml:38]: No criticism, just a hint 😊

File: tests/test_mocks.py
[the remark this answers is not in the document]
Comment (GitOldGrumpy) [reply tests/test_mocks.py:1]: Re-ordered.

File: tests/test_mocks.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+def remove_cache() -> None:  # noqa: D103
+    # Remove cache file if exists
+    try:
+        remove(CACHE_FILENAME)
+    except FileNotFoundError:
+        pass
+
+
+def build_routes(httpx_mock, filenames: List[str]) -> None:  # noqa: D103
```
[the remark this answers is not in the document]
Comment (GitOldGrumpy) [reply tests/test_mocks.py:24]: If you could give me some pointers on this one? I worked out how to turn remove_cache into a fixture maybe incorrectly, so you might want to check that. I cant seem to apply this to the convert this to a fixture. I need access to httpx_mock which is provided by another fixture.

File: tests/test_mocks.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+    client = MyT("user@email.info", "password")
+    # Nothing validates this is correct, just replays a refresh token sequence
+    await client.login()
+
+
+@pytest.mark.asyncio
+async def test_get_static_data(httpx_mock: HTTPXMock):  # noqa: D103
+    #  Create valid token => Means no authentication requests
+    with open("./data/mocks/cached_token.json") as f:  # noqa: ASYNC101. I dont see this as a problem in test code
```
[the remark this answers is not in the document]
Comment (GitOldGrumpy) [reply tests/test_mocks.py:87]: Done.

**Comment by GitOldGrumpy** [comment]
@CM000n you might want to double check my attempts to eradicate 3.8 from building. I was also not sure if we need to update the poetry file so that `python-versions ` on the `[[package]]` are now also 3.9?

File: mytoyota/controller.py [hunk 4]
```diff
@@ -105,6 +105,8 @@ async def _authenticate(self):
                     self._authenticate_url, json=data
                 )  # , headers=standard_headers)
                 _LOGGER.debug(format_httpx_response(resp))
+                #                print(f"{format_httpx_response_json(resp)},", flush=True)
```
Comment (CM000n) [inline mytoyota/controller.py:108]: This commented out print statement should be removed before the merge.

File: mytoyota/controller.py [hunk 5]
```diff
@@ -123,6 +125,8 @@ async def _authenticate(self):
                 headers={"cookie": f"iPlanetDirectoryPro={data['tokenId']}"},
             )
             _LOGGER.debug(format_httpx_response(resp))
+            #            print(f"{format_httpx_response_json(resp)},", flush=True)
```
Comment (CM000n) [inline mytoyota/controller.py:128]: This commented out print statement should be removed before the merge.

File: mytoyota/controller.py [hunk 6]
```diff
@@ -142,6 +146,8 @@ async def _authenticate(self):
                 },
             )
             _LOGGER.debug(format_httpx_response(resp))
+            #            print(f"{format_httpx_response_json(resp)},", flush=True)
```
Comment (CM000n) [inline mytoyota/controller.py:149]: This commented out print statement should be removed before the merge.

File: mytoyota/controller.py [hunk 7]
```diff
@@ -168,6 +174,8 @@ async def _refresh_tokens(self) -> None:
...
```
Comment (CM000n) [inline mytoyota/controller.py:177]: This commented out print statement should be removed before the merge.

File: mytoyota/controller.py [hunk 8]
```diff
@@ -254,6 +262,8 @@ async def request_raw(  # noqa: PLR0913
                 follow_redirects=True,
             )
             _LOGGER.debug(format_httpx_response(response))
+            #            print(f"{format_httpx_response_json(response)},", flush=True)
```
Comment (CM000n) [inline mytoyota/controller.py:265]: This commented out print statement should be removed before the merge.

File: tests/integration_tests/conftest.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+import pytest
+
+from mytoyota.controller import CACHE_FILENAME
+
+
+@pytest.fixture
+def data_folder(request) -> str:
+    """Return the folder containing test files."""
+    return f"{Path(request.module.__file__).parent}/data"
```
Comment (CM000n) [inline tests/integration_tests/conftest.py:13]: You don't need an f-string here:
`str(Path(request.module.__file__).parent / "data")`

File: tests/unit_tests/conftest.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+"""Need for pytest or else it will cause an import error in pytest."""
+from pathlib import Path
+
+import pytest
+
+
+@pytest.fixture
+def data_folder(request) -> str:
+    """Return the folder containing test files."""
+    return f"{Path(request.module.__file__).parent}/data"
```
Comment (CM000n) [inline tests/unit_tests/conftest.py:10]: You don't need an f-string here:
`str(Path(request.module.__file__).parent / "data")`

File: tests/unit_tests/conftest.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+"""Need for pytest or else it will cause an import error in pytest."""
+from pathlib import Path
+
+import pytest
+
+
+@pytest.fixture
```
Comment (CM000n) [inline tests/unit_tests/conftest.py:7]: I think the scope for this ficture can be at least `module`, right?

File: tests/integration_tests/conftest.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
+"""Need for pytest or else it will cause an import error in pytest."""
+
+from pathlib import Path
+
+import pytest
+
+from mytoyota.controller import CACHE_FILENAME
+
+
+@pytest.fixture
```
Comment (CM000n) [inline tests/integration_tests/conftest.py:10]: I think the scope for this ficture can be at least `module`, right?

**Review by CM000n — CHANGES_REQUESTED** [review]

**Comment by CM000n** [comment]
> @CM000n you might want to double check my attempts to eradicate 3.8 from building. I was also not sure if we need to update the poetry file so that `python-versions ` on the `[[package]]` are now also 3.9?

Looks good to me so far 😊  You should only also remove the entry for Python 3.8 in the `classifiers` in the `pyproject.toml` file

**Comment by CM000n** [comment]
Mhmm, doesn't look like @joro75 is particularly active. I actually don't want to wait too long as these changes are very beneficial in terms of test coverage.

What do you think @DurgNomis-drol? Should we give ourselves a specific timeframe here for feedback from @joro75 and merge the changes otherwise?

**Comment by joro75** [comment]
Hi all,

Sorry for the delay. I have been busy, and have not been able to follow all the recent changes.

Dropping Python 3.8 is OK for me. I also plan to update Domoticz-Toyota-Plugin to Python 3.9 soon, also updating it to the latest mytoyota. 
It is correct that the original Domoticz documentation regarding python is old. Using Python >= 3.9 in combination with Python is possible, as I recently updated my own installation to Python 3.9.

Just drop Python 3.8, and continue with the merge!

**Comment by CM000n** [comment]
Thanks for you response @joro75 😊

**Review by CM000n — APPROVED** [review]

**Pull request merged**